tests/conftest.py

```diff
@@ -170,13 +178,39 @@ def _generate_calculation_node(process_state=ProcessState.FINISHED, exit_status=
         if process_state is ProcessState.FINISHED and exit_status is None:
             exit_status = 0
 
-        node = CalculationNode(process_type=entry_point)
-        node.set_process_state(process_state)
+        calculation_node = CalculationNode(process_type=entry_point)
+        calculation_node.set_process_state(process_state)
 
         if exit_status is not None:
-            node.set_exit_status(exit_status)
+            calculation_node.set_exit_status(exit_status)
+
+        if repository is not None:
+            # Or use from_tree?
+            calculation_node.base.repository.put_object_from_filelike(repository[0], repository[1])
+
+        # For storing, need to first store the input nodes, then the CalculationNode, then the output nodes
+        if inputs is not None:
+            for input_label, input_node in inputs.items():
+                calculation_node.base.links.add_incoming(
+                    input_node,
+                    link_type=LinkType.INPUT_CALC,
+                    link_label=input_label,
+                )
+
+                input_node.store()
+
+        if outputs is not None:
+            # Need to first store CalculationNode before I can attach `created` outputs
+            calculation_node.store()
```

Comment:
I think you want to move this line outside of this conditional. Probably just before it, after the block that added the inputs. The input links are not "flushed" to the database until the calc node is stored itself.

Reply:
The same reason here as in my comment below. If I always store it, I'm getting a:
```python
ValueError: attempted to add an input link after the process node was already stored.
```
in `validate_incoming` when constructing a nested workchain for testing.

Reply:
Let's not hold up this PR over this issue. Let's just keep it in the outputs conditional so we can merge this.

Reply:
Alright, so if it's fine that we keep it as it is, then this should be ready for merge, @sphuber. I have just removed the underscores from the `_dump_node_yaml`-internal variables, as suggested by @agoscinski in my latest commit.